File: 2294. Partition Array Such That Maximum Difference Is K/partition_array_such_that_maximum_difference_is_k.py
class Solution:
    def partitionArray(self, nums: List[int], k: int) -> int:
        nums.sort()
        res=0
        cur=nums[0]
        for n in nums:
            if(n-cur)>k:
                res+=1
                cur=n
        return res+1

# Time complexity:
# O(nlogn)

# Space complexity:
# O(1)

File: 2294. Partition Array Such That Maximum Difference Is K/partition_array_such_that_maximum_difference_is_k.cpp
class Solution {
public:
    int partitionArray(vector<int>& nums, int k) {
        sort(nums.begin(),nums.end());
        int n=nums.size();
        //result will get stored in subsequences
        int subsequences = 0;
        int min_val = nums[0];
        for(int i=1;i<n;++i){
            if(nums[i]-min_val > k){
                min_val = nums[i];
                subsequences++;
            }
        }
        return subsequences+1;  //last one didnt get count
    }
};

// Time complexity:
// O(nlogn)

// Space complexity:
// O(1)
